Require TagLib 1.11 or newer

uklotzde opened the pull request:
...to get rid of untested legacy code.

Do we need a redundant check in CMakeLists.txt in addition to the `static_assert` in the code? I have left a TODO comment.

This might also fix https://bugs.launchpad.net/bugs/1940777: The invocation of `SoundSource::importTrackMetadataAndCoverImage()` in `SoundSourceOpus::importTrackMetadataAndCoverImage()` was accepted by all compilers even though this is slightly misleading. It should have been `MetadataSource::importTrackMetadataAndCoverImage()`.

uklotzde committed:
Require TagLib 1.11 or newer

Be-ing commented on CMakeLists.txt:
```diff
@@ -2366,6 +2366,9 @@ else()
 endif()
 
 # TagLib
+# TODO: Check for minimum required version?
+# See also: static_assert in src/track/taglib/trackmetadata_common.h
+# that verifies TAGLIB_MAJOR_VERSION/TAGLIB_MINOR_VERSION.
```
I would expect a failure at configure time rather than build time for an outdated dependency, so I think it would be better to check the minimum version with CMake than static_assert.

uklotzde replied:
We can replace the static_assert with a configure check if someone provides it. Until then the static_assert is even more safe although less convenient.

Be-ing replied:
How is it more safe?

uklotzde replied:
Nearer at the code. Let's agree on just safe ;) We also have other version checks in the code.

But it would be preferable to check the minimum required version beforehand.

Be-ing replied:
> if someone provides it

Done: https://github.com/uklotzde/mixxx/pull/17

Be-ing replied:
In my experience, it is an unpleasant surprise to have a build fail because of an outdated dependency. I prefer to have the build system tell me up front that it will not work and why.

Be-ing replied:
Also I think it is easier for us to keep track of what our minimum required dependencies are in the build system than having to dig in the C++ code for that information.

Be-ing commented:
Ubuntu 20.04 [ships TagLib 1.11.1](https://packages.ubuntu.com/focal/libtag1-dev). To maintain adherence to our [minimum requirements policy](https://github.com/mixxxdj/mixxx/wiki/Minimum%20requirements%20policy) we need to either:
  * Wait on removing these hacks for old TagLib versions.
  * Vendor TagLib 1.12
  * Tell developers on Ubuntu 20.04 to build TagLib from source or vcpkg. This could be a decent option if we add a vcpkg.json to this repository.
  * Tell developers on Ubuntu 20.04 to build a Flatpak manifest. GNOME Builder has nice integration with this, but I don't know of other IDEs that do.

Be-ing commented:
Disregard the last comment. I mistakenly thought TagLib 1.12 was required.

uklotzde commented:
This PR does not conflict with the minimum requirements policy. We only require TagLib 1.11.x here. This version has serious, known bugs (e.g. OGG corruption) and we still maintain workarounds.

Be-ing committed:
CMake: check for TagLib >= 1.11

Be-ing commented:
+38 −228 :)

uklotzde commented:
Why does the build fail on Windows? We package version 1.12.

Be-ing commented:
Are you sure the current Windows build environment has TagLib 1.12? This might need to wait for #4225.

Be-ing commented:
Oh I think it is an issue with the version detection.

Be-ing committed:
CMake: fix version detection for TagLib

On Windows, the pkgconfig file is found and the variable is
defined but somehow set to empty string.

Be-ing commented:
I fixed the version detection on Windows: https://github.com/uklotzde/mixxx/pull/18

uklotzde committed:
Merge pull request #18 from Be-ing/taglib_cmake_version_check

CMake: fix version detection for TagLib
